### PR TITLE
Display in query when user disconnects and re-connects using ISON, closes #337

### LINKED ISSUE #337: User join/leave/disconnect not shown in query window
I suggest query windows should show when the user disconnects or joins one of the channel we're both on.

**Example:**
As a persistent service, a user is often away form the IRC window for a long time, thus when another user opens a query with the user, it's good to know whether that other user is still online or not when eventually getting the message.

**Implications:**
These notifications are only caught when join/leave/disconnect events for a user happen in a channel we are also in. Kinda hackish.

Suggestions (taken from conversation about that issue on #thelounge irc channel):
- Using the topic bar to display something like "Online on: #thelounge #freenode" or "Last seen 20 minutes ago on #thelounge"( @maxpoulin64 )  , add hostmask as well ( @xPaw )

### PR DESCRIPTION
**Pull request by MaxLeiter**
WIP, almost done.

Need a better way to create the messages - ideally they'd be created on the server (so they could be assigned a msgid), however due to the nature of `ison` that's a PITA. The lack of an ID causes the unread marker to act a little odd. @xPaw, suggestions?

Also, should somehow compare hostnames between when the user was last connected and if they are now. Again, due to `ison` primitiveness, not sure best way to do this. Suggestions appreciated.

screen shot 2017-11-11 at 12 07 23 am

### CODE REVIEW

File: client/js/lounge.js [hunk 1]
```diff
@@ -312,6 +312,15 @@ $(function() {
 		if (!target) {
 			return;
 		}
+
+		const targetTitle = self.data("title");
+		if (targetTitle.indexOf("#") === -1 && !self.hasClass("lobby")) { // no # or .lobby; assume it's a user query
```
Comment (xPaw) [inline client/js/lounge.js]: This entire thing can be done on the server inside `open`, and it can check if open channel type is query.

**Comment by xPaw** [comment]
I don't see a reason why all this code is on the client, you say it's a PITA to do on the server, but what's the reason because I fail to see one. You can add `state` (or something like that) on `Chan` object on the server to send updates to client correctly.

I would also modify `quit` event to show quits for common channels in query.

File: client/js/constants.js [hunk 2]
```diff
@@ -72,6 +72,8 @@ const actionTypes = [
 	"away",
 	"back",
 	"ban_list",
+	"connect",
```
Comment (xPaw) [inline client/js/constants.js]: Reusing quit/join could probably work just fine.

File: src/plugins/irc-events/ison.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+module.exports = function(irc) {
+	const client = this;
+	irc.on("users online", (data) => {
+		if (data.nicks.length === 1) {
+			const target = client.find(client.lastActiveChannel);
+			const chan = target.chan;
+			if (chan.type === Chan.Type.QUERY) {
+				const messages = chan.messages;
+				let lastMessageType;
+				for (let i = 0; i < (messages.length > 10 ? 10 : messages.length - 1); i++) {
```
Comment (xPaw) [inline src/plugins/irc-events/ison.js]: Why not simply add `isOnline` to `chan` object and compare that?

File: src/plugins/irc-events/ison.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
+"use strict";
+
+const Msg = require("../../models/msg");
+const Chan = require("../../models/chan");
+
+module.exports = function(irc) {
+	const client = this;
+	irc.on("users online", (data) => {
+		if (data.nicks.length === 1) {
+			const target = client.find(client.lastActiveChannel);
```
Comment (xPaw) [inline src/plugins/irc-events/ison.js]: This is gonna be weird with multiple clients open. As we discussed on IRC, the better way is most likely querying all names with `ison` at once.

Comment (MaxLeiter) [reply src/plugins/irc-events/ison.js]: and just compare the array of open queries with the returned array and find what's missing?

Comment (xPaw) [reply src/plugins/irc-events/ison.js]: Pretty much, yeah.

**Comment by astorije** [comment]
Hey @MaxLeiter, do you think you'll get back to this? :)

**Comment by MaxLeiter** [comment]
yeah, want to finish up nicks in links first

**Comment by astorije** [comment]
Cool! It seems to me that this other PR is almost done, so that's great!

**Comment by xPaw** [comment]
@MaxLeiter want to revisit?

**Comment by astorije** [comment]
@MaxLeiter, @xPaw, do you know what's the status of this?

**Comment by MaxLeiter** [comment]
@astorije: requires a full rewrite on how it works. It's going to have an array of users + their status and refresh everyone n minutes (or something similar), instead of when you click on the query.

At least, that's what I remember. @xPaw?

**Comment by xPaw** [comment]
I believe we talked about using MONITOR for this and as a fallback use quit/part/join messages (and probably ISON on open like it is now).

And instead of iterating last X messages use `Chan.State` to store whether user is currently online.

**Comment by astorije** [comment]
@MaxLeiter, it looks like the pr auto closed, you can reopen it by pushing new commits to the branch

**Comment by MaxLeiter** [comment]
Yeah I’m working locally before commiting.

**Pull request closed without merging**